crypto/default/src/main/java/org/keycloak/crypto/hash/Argon2PasswordHashProviderFactory.java

```diff
@@ -23,6 +24,13 @@ public class Argon2PasswordHashProviderFactory implements PasswordHashProviderFa
     public static final String ITERATIONS_KEY = "iterations";
     public static final String PARALLELISM_KEY = "parallelism";
 
+    /**
+     * The Argon2 password hashing is CPU bound, so it doesn't make sense to hash more values concurrently than there are cores on the machine.
+     * When we run more, this only leads to an increased memory usage and to throttling of the process in containerized environments
```

Comment:
Are users going to adjust the memory parameter of Argon2 (for example going up to 46MiB to use a single iteration)? That coupled with running on a node with a lot of cores, like a m5d.24xlarge, with a high or no cpu limit could still be quite a lot of memory.

Reply:
I think we should reflect that in docs; and also make the number of parallell hash operations configurable.

From what I found though increasing memory doesn't decrease execution time (at least not on Java with BouncyCastle implementation), so best practice would probably be to use the min memory recommended by OWASP.